tt_metal/impl/allocator/algorithms/free_list_opt.cpp

```diff
@@ -0,0 +1,607 @@
+    // 128MB as the last seggregated class size should be enough
+    // avoid having too many classes as iterating them is not free
+    ssize_t max_count = intlg2(128 * 1024 * 1024 / size_segregated_base);
+    return std::clamp(count, ssize_t{2}, max_count);
+}
+
+namespace tt {
+
+namespace tt_metal {
+
+namespace allocator {
+
+FreeListOpt::FreeListOpt(
+    DeviceAddr max_size_bytes, DeviceAddr offset_bytes, DeviceAddr min_allocation_size, DeviceAddr alignment) :
+    size_segregated_count((num_segerated_classes(max_size_bytes, size_segregated_base))),
+    Algorithm(max_size_bytes, offset_bytes, min_allocation_size, alignment) {
+    // Reduce reallocations by reserving memory for free list components
+    constexpr size_t initial_block_count = 64;
+    block_address_.reserve(initial_block_count);
+    block_size_.reserve(initial_block_count);
+    block_prev_block_.reserve(initial_block_count);
+    block_next_block_.reserve(initial_block_count);
+    block_is_allocated_.reserve(initial_block_count);
+    free_meta_block_indices_.reserve(initial_block_count);
+    meta_block_is_allocated_.reserve(initial_block_count);
+    free_blocks_segregated_by_size_.resize(size_segregated_count);
+    for (auto& free_blocks : free_blocks_segregated_by_size_) {
+        free_blocks.reserve(initial_block_count);
+    }
+    allocated_block_table_.resize(n_alloc_table_buckets);
+    for (auto& bucket : allocated_block_table_) {
+        bucket.reserve(n_alloc_table_init_bucket_size);
+    }
+
+    init();
+}
+
+void FreeListOpt::init() {
+    max_size_bytes_ += shrink_size_;
+    shrink_size_ = 0;
+
+    block_address_.clear();
+    block_size_.clear();
+    block_prev_block_.clear();
+    block_next_block_.clear();
+    block_is_allocated_.clear();
+    free_meta_block_indices_.clear();
+    meta_block_is_allocated_.clear();
+    for (auto& bucket : allocated_block_table_) {
+        bucket.clear();
+    }
+    for (auto& free_blocks : free_blocks_segregated_by_size_) {
+        free_blocks.clear();
+    }
+
+    // Create a single block that spans the entire memory
+    block_address_.push_back(0);
+    block_size_.push_back(max_size_bytes_);
+    block_prev_block_.push_back(-1);
+    block_next_block_.push_back(-1);
+    block_is_allocated_.push_back(false);
+    meta_block_is_allocated_.push_back(true);
+    free_blocks_segregated_by_size_[get_size_segregated_index(max_size_bytes_)].push_back(0);
+}
+
+std::optional<DeviceAddr> FreeListOpt::allocate(DeviceAddr size_bytes, bool bottom_up, DeviceAddr address_limit) {
+    DeviceAddr alloc_size = align(std::max(size_bytes, min_allocation_size_));
+
+    // Find the best free block by looking at the segregated free blocks, if we can find a block in it's size class
+    // we can be confident that it's the best block to allocate from. Else, look at the next size class. However the
+    // blocks within a size class are not sorted by size, so we may not always find the best block.
```

Comment:
While best fit is a better policy for fragmentation, I think the use of it is causing [CB < -> L1 buffer crashes](https://github.com/tenstorrent/tt-metal/actions/runs/12332336779/job/34421152920). 

Currently we allocate L1 buffers first fit top down and for each program CBs are allocated bottom up. With this change we could ignore large size classes of free regions and then allocate a buffer at a lower addr space. This could contend with the CBs. 

Could this check be uplifted to always pick the free size class that is at higher addr space? @marty1885

Reply:
@abhullar-tt and @marty1885  does our current allocator require the following specialization :

* FirstMatch (ie first fit) top-down for L1 buffers
* FirstMatch bottom-up for CBs local to the Program?

Reply:
yeah for L1 buffers this is the case to avoid crashing into CBs.

Reply:
* L1 buffers, top-down, is analogous to heap (longer term memory, lifetime across OPs)
* CBs, bottom-up, is analogous to stack (shorter term memory, lifetime within an OP)

Reply:
yeah, more details on when CBs [here](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/memory/allocator.md#3-sram-banks-and-circularbuffers)

Reply:
How about we keep using the current (slower) FreeList allocator for L1 and use my optimized one for DRAM? 

I did some research and figured that the segregated approach would incur significant overhead to support first fit. The new scanning algorithm would have to be implemented as a different class to keep performance.

Reply:
yes, that should work. Ill test out that change on CI

Reply:
It should be noted that we also allocate buffers in different directions in dram. Regular/user buffers are allocated bottom up in dram, but program binaries are allocated top down since these are basically persistent, so we allocate top down to avoid fragmenting the space of user buffers.

Reply:
Oops.. the comment is outdated. The code now always finds the smallest block now as it is sorted. I'll delete the comment in code.

Reply:
hmm @tt-dma do you have the dispatch data collection results and does it have sizes of programs? in the general case if programs are around the same size would they just fit in the same section and since this segregated list has better fragmentation we might not run into issues with that. wdyt @tt-aho

Reply:
For context, the segregated list approach is FreeList + BestFit. I think the true question is why TT was using FirstFit.

Reply:
FirstFit allowed us to ensure that allocating space from different directions (top-down / bottom-up) did not end up in the same place

Reply:
I added some code so it defaults to FirstFit now. Please run all tests again